llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -772,6 +773,66 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
           unsigned CurrentOffset = Ins.first->second;
           if (IsNewSpecConstant) {
             unsigned Size = M.getDataLayout().getTypeStoreSize(SCTy);
+            unsigned Align = M.getDataLayout().getABITypeAlignment(SCTy);
+
+            // Ensure correct alignment
+            if (CurrentOffset % Align != 0) {
+              // Compute necessary padding to correctly align the constant.
+              Padding = Size - CurrentOffset % Align;
+
+              // Update offsets.
+              NextOffset += Padding;
+              CurrentOffset += Padding;
+              OffsetMap[SymID] = NextOffset;
+
+              // The spec constant map can't be empty as the first offset is 0
+              // and so it can't be misaligned.
+              assert(!SCMetadata.empty() &&
```

Comment:
I would also suggest to add assert(CurrentOffset % Align == 0, "alignment calculation error");